src/menu/charprofile.c

```diff
@@ -0,0 +1,368 @@
+
+	r_color4(0, 0, 0, 0.5);
+	r_shader_standard_notex();
+	r_mat_mv_translate(-150, descbg_ofs + m->drawdata[2] * 0.5, 0);
+	r_mat_mv_scale(650, m->drawdata[2], 1);
+	r_draw_quad();
+	r_shader_standard();
+	r_mat_mv_pop();
+
+	CharProfiles i = ctx->char_draw_order[m->cursor];
+
+	MenuEntry *e = dynarray_get_ptr(&m->entries, m->cursor);
+
+	float o = 1 - e->drawdata*2;
+	float pbrightness = 0.6 + 0.4 * o;
+
+	float pofs = fmax(0.0f, e->drawdata * 1.5f - 0.5f);
+	pofs = glm_ease_back_in(pofs);
+
+	if(i != m->selected) {
+		pofs = lerp(pofs, 1, menu_fade(m));
+	}
+
+	int selected = check_unlocked_profile(m->cursor);
+
+	if(selected != LOCKED_PROFILE) {
+		Sprite *spr = profiles[selected].sprite;
+		SpriteParams portrait_params = {
+			.pos = { SCREEN_W/2 + 240 + 320 * pofs, SCREEN_H - spr->h * 0.5 },
+			.sprite_ptr = spr,
+			.shader = "sprite_default",
+			.color = RGBA(pbrightness, pbrightness, pbrightness, 1),
+		};
+
+		r_draw_sprite(&portrait_params);
+		portrait_params.sprite_ptr = portrait_get_face_sprite(profiles[selected].name, profiles[selected].faces[ctx->face]);
+		r_draw_sprite(&portrait_params);
+	}
+	r_mat_mv_push();
+	r_mat_mv_translate(SCREEN_W/4, SCREEN_H/3, 0);
+	r_mat_mv_push();
+	r_mat_mv_push();
+
+	if(e->drawdata != 0) {
+		r_mat_mv_translate(0, -300 * e->drawdata, 0);
+		r_mat_mv_rotate(M_PI * e->drawdata, 1, 0, 0);
+	}
+
+	text_draw(profiles[selected].fullname, &(TextParams) {
+		.align = ALIGN_CENTER,
+		.font = "big",
+		.shader = "text_default",
+		.color = RGBA(o, o, o, o),
+	});
+	r_mat_mv_pop();
+
+	if(e->drawdata) {
+		o = 1 - e->drawdata * 3;
+	} else {
+		o = 1;
+	}
+
+	text_draw(profiles[selected].title, &(TextParams) {
+		.align = ALIGN_CENTER,
+		.pos = { 20*(1-o), 30 },
+		.shader = "text_default",
+		.color = RGBA(o, o, o, o),
+	});
+	r_mat_mv_pop();
+
+	r_color4(1, 1, 1, 0.5);
+	text_draw_wrapped(profiles[selected].description, DESCRIPTION_WIDTH, &(TextParams) {
+		.align = ALIGN_LEFT,
+		.pos = { -190, 120 },
+		.font = "small",
+		.shader = "text_default",
+		.color = RGBA(o, o, o, o),
+	});
+	r_mat_mv_pop();
+
+	o = 0.3*sin(m->frames/20.0)+0.5;
+	r_shader("sprite_default");
+
+	r_draw_sprite(&(SpriteParams) {
+		.sprite = "menu/arrow",
+		.pos = { 30, SCREEN_H/3+10 },
+		.color = RGBA(o, o, o, o),
+		.scale = { 0.5, 0.7 },
+	});
+
+	r_draw_sprite(&(SpriteParams) {
+		.sprite = "menu/arrow",
+		.pos = { 30 + 340, SCREEN_H/3+10 },
+		.color = RGBA(o, o, o, o),
+		.scale = { 0.5, 0.7 },
+		.flip.x = true,
+	});
+
+	r_state_pop();
+}
+
+static void action_show_character(MenuData *m, void *arg) {
+	return;
+}
+
+static void add_character(MenuData *m, int i) {
+	if(strcmp(profiles[i].name, "locked")) {
+		log_debug("adding character: %s", profiles[i].name);
+		portrait_preload_base_sprite(profiles[i].name, NULL, RESF_PERMANENT);
+		profiles[i].sprite = portrait_get_base_sprite(profiles[i].name, NULL);
+		MenuEntry *e = add_menu_entry(m, NULL, action_show_character, NULL);
```

Comment:
You could use `e->arg` to store the sprite instead of reserving space for pointers in the global array. You can also declare the profiles array `const` then.